Add Conda Lock Updating Action

ajelinski opened the pull request:
This GitHub Action creates or updates Conda Lock from the freshly created Conda Environment based on the Conda `environment.yml` file. The Pull Request is automatically issued if there is any change in the Conda Lock.

The workflow using this Action is available at: https://github.com/antmicro/symbiflow-examples-bot/blob/master/.github/workflows/update_conda_locks.yml
The example workflow jobs can be checked at: https://github.com/antmicro/symbiflow-examples-bot/actions/runs/453502941
The lock is quite fresh in there on the `master` branch so apart from the environment's name change (due to the script change) in the latest PR there's only one package updated but it can be seen how the Conda Lock looks like: https://github.com/antmicro/symbiflow-examples-bot/pull/41/files

umarcor commented on update_conda_lock/action.yml:
```diff
@@ -0,0 +1,129 @@
+inputs:
+  commit_message:
+    description: 'Message of the bump commit'
+    required: true
```
I'm not sure about this being check by the backend. I suggest to try it, and check it explicitly in the script instead.

umarcor commented on update_conda_lock/action.yml:
```diff
@@ -0,0 +1,129 @@
+  user_email:
+    description: 'Email address of the user creating commit'
+    required: true
+    default: '<>'
```
Instead of providing a dummy default, I suggest not providing a default.

ajelinski replied:
It seems to me that without providing any default value using this action in a workflow would always require passing this value. This dummy default simply serves the purpose of satisfying the command `git config user.email` if the workflow didn't pass that value to the Action.

umarcor replied:
@ajelinski, I believe that's the purpose of `required`. If it is required, the content needs to be meaningful. If the content can be irrelevant, then it should be `required: false` and the Action should check it and provide a default internally.

umarcor commented on update_conda_lock/action.yml:
```diff
@@ -0,0 +1,129 @@
+  - id: lock-update
+    shell: bash
+    run: |
+      echo "::group::Check git"
```
Please, add this to a shell script and call it from here.

Moreover, it seems that most of it is not specific to the action, but general info about the environment. Therefore, I suggest to create two scripts.

umarcor commented on update_conda_lock/action.yml:
```diff
@@ -0,0 +1,129 @@
+      echo "::endgroup::"
+
+      echo "::group::Add And Commit Changes"
+      set -x
```
I would wrap the grouping in a couple of helper functions:

```
gstart() {
  echo "::group::$1"
  set -x
}

gend() {
   set +x
   echo "::endgroup::"
}
```

ajelinski commented:
I can see that I need to add `#.*coding: utf-8` and LICENSE to satisfy the `Checks` test. Should the license simply resemble that found in the other files @mithro @umarcor? https://github.com/SymbiFlow/actions/blob/77e0cafc07a9dfbe39d41aad30c57c0211ba5af6/checks/check_license.sh#L3-L7

umarcor commented on update_conda_lock/action.yml:
```diff
@@ -0,0 +1,129 @@
+
+      echo "::group::Create a Pull Request"
+      # Export variables used by the script
+      export GITHUB_TOKEN="${{ inputs.github_token }}"
```
```
  # PR_BASE has been already exported before the `git checkout -b` command
  GITHUB_TOKEN="${{ inputs.github_token }}" \
  GITHUB_REPOSITORY="${{ github.repository }}" \
  PR_HEAD="$BRANCH_NAME" \
  PR_TITLE="${{ inputs.pr_title }} $(date -u +'%Y-%m-%d %H:%M %p %Z')" \
  python3 $GITHUB_ACTION_PATH/create_pull_request.py
```

However, since you are doing most of the logic in Python, why don't you include this in there too? All input parameters should be available as environment variables. Check `env` in bash or `os.environ` in Python.

umarcor commented:
> I can see that I need to add #.*coding: utf-8 and LICENSE to satisfy the Checks test. Should the license simply resemble that found in the other files @mithro @umarcor? 

Yes, please. Unless you have any specific requirement.

umarcor commented:
Please, add it to the README too.

ajelinski commented:
I had to make some small fixes after testing on `symbiflow-arch-defs` CI. I've fixed failing checks although I have to make sure about the copyright. Therefore I'll make it draft for now.

mithro commented:
@ajelinski - Ping? I would like to get this GitHub action moving again?

ajelinski commented:
I still need to test if everything works well after these changes but please take a look if you like it more now @umarcor.

ajelinski commented:
Rebased and fixed after testing on a `symbiflow-examples` repository clone. A PR created by this action is e.g.
https://github.com/antmicro/symbiflow-examples-bot/pull/142 (created during https://github.com/antmicro/symbiflow-examples-bot/actions/runs/797726587).

umarcor reviewed (approved):
The structure looks much nicer and easier to understand/maintain. Congrats for the cleanup!

ajelinski commented:
I'm glad you like it @umarcor!

I've cleaned the commit history. I think now this PR is ready to be merged.

ajelinski commented:
I tested the bot again to make sure that everything still works as intended before it's merged.

Somehow after separating scripts and steps I skipped the test of the special case when the lock is up to date. It needed a little modification to work properly (`exit 0` was enough when it was a single script): https://github.com/SymbiFlow/actions/compare/d358d22afdaf54376197be3bd1c9af265e79ecb2..6970a7e6257a3e66dec67b97cd51c44a0c34ad5b

BTW there's also an issue with pip that doesn't handle simultaneous installation of git-based packages with their git-based  dependencies properly. I've found a way to avoid it, i.e. making pip ignore package dependencies when Conda creates the environment based on the Conda Lock. It is absolutely safe to do since all packages, including the dependencies, are in the Conda Lock anyway.

I have created an issue on pip so maybe it won't be needed in the future: https://github.com/pypa/pip/issues/10002

For now, I added information on how to handle such a situation to the README: https://github.com/SymbiFlow/actions/pull/5/commits/eec0e6465f49e2e57828ff379e0fd162e0980540 .

Could you please take another look at the action changes @umarcor? Now the action is definitely ready to be merged if you find them acceptable.

umarcor commented:
LGTM! @mithro, I'll let you merge :wink:

mithro commented:
> BTW there's also an issue with pip that doesn't handle simultaneous installation of git-based packages with their git-based dependencies properly. I've found a way to avoid it, i.e. making pip ignore package dependencies when Conda creates the environment based on the Conda Lock. It is absolutely safe to do since all packages, including the dependencies, are in the Conda Lock anyway.
> 
> I have created an issue on pip so maybe it won't be needed in the future: [pypa/pip#10002](https://github.com/pypa/pip/issues/10002)

We should **not** have any packages which use the `@` symbol in their `install_requires` as we determined that functionality is fundamentally broken. Instead the `install_requires` should list the package name and then the `requirements.txt` should specify how/where to get the package from (IE a `git+http://` URL).

mithro commented on README.md:
```diff
@@ -6,6 +6,44 @@ This repository contains [actions](https://docs.github.com/en/free-pro-team@late
 
 Runs license and python linting checks.
 
+✔️ [symbiflow/actions/update_conda_lock](update_conda_lock)
```
This should be a header.

mithro commented on README.md:
```diff
@@ -6,6 +6,44 @@ This repository contains [actions](https://docs.github.com/en/free-pro-team@late
+Running this action creates a Pull Request that creates or updates a Conda Lock.
+Checking workflow results and merging the Pull Request are the only actions that require repository maintainer's attention.
+
+Please bear in mind that the Pull Request can only trigger workflows if this action is used with GitHub's [Personal Access Token](https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token).
```
FYI - A deploy keys with push access is a better solution than PATs as they only give you access to a single repository.

ajelinski replied:
It's a nice option!

mithro commented on README.md:
````diff
@@ -6,6 +6,44 @@ This repository contains [actions](https://docs.github.com/en/free-pro-team@late
+```
+In case passing *PAT* is not an option, passing `secrets.GITHUB_TOKEN` will also result in creating a Pull Request but without triggering PR Checks.
+
+The inputs to this action are (bold ones are **required**):
````
What are the default values for the non-required arguments?

mithro commented on update_conda_lock/action.yml:
```diff
@@ -0,0 +1,75 @@
+    run:   $GITHUB_ACTION_PATH/check_prepare_env.sh
+
+  - shell: bash
+    run: |
```
This would probably be better as a Python script?

ajelinski replied:
I'm not sure if you meant it this way but I can't make it a separate script since it uses GH Action inputs so it has to be in the `action.yml` file.

If you just thought about converting this step to Python and leaving it in `action.yml` then I'm not sure what are the advantages?

mithro commented on update_conda_lock/check_prepare_env.sh:
```diff
@@ -0,0 +1,48 @@
+#!/bin/bash
```
If we make this file Python it can work on windows / mac / linux? We already require Python for other parts.

ajelinski replied:
This script should work on all these OSs, there's nothing really OS-specific.

mithro commented on update_conda_lock/push_updated_lock.sh:
```diff
@@ -0,0 +1,50 @@
+#!/bin/bash
```
If we make this file Python it can work on windows / mac / linux? We already require Python for other parts?

ajelinski replied:
The script itself should work on all these OSs but I'm not sure about `update_lock.py` on Windows with its `\r\n` newlines as there are some file manipulations involved.

mithro commented on update_conda_lock/create_pull_request.py:
```diff
@@ -0,0 +1,61 @@
+#!/usr/bin/env python3
+#.*coding: utf-8
```
```suggestion
# -*- coding: utf-8 -*-
```

mithro commented on update_conda_lock/update_lock.py:
```diff
@@ -0,0 +1,257 @@
+#!/usr/bin/env python3
+#.*coding: utf-8
```
```suggestion
# -*- coding: utf-8 -*-
```

mithro commented on update_conda_lock/update_lock.py:
```diff
@@ -0,0 +1,257 @@
+
+        # Find core of the dependency line without version etc.
+        core_match = search(r'(^|\s)([^\s-][^\s=<>~!;]+)', only_dependency)
+        if core_match is not None:
```
The indenting here is getting pretty extreme.
```suggestion
        if core_match is None:
            continue
```

ajelinski replied:
Thanks for pointing that out.

mithro commented on update_conda_lock/update_lock.py:
```diff
@@ -0,0 +1,257 @@
+    print('* ' + env_name + ': ' + env_var)
+    return env_var
+
+# Returns True if lock file has been updated, False otherwise
```
This comment should be the docstring not a commant.

mithro reviewed (approved):
Some minor formatting / style issues.

Can you run pylint or flake8 on your Python files?

ajelinski commented:
> > BTW there's also an issue with pip that doesn't handle simultaneous installation of git-based packages with their git-based dependencies properly. I've found a way to avoid it, i.e. making pip ignore package dependencies when Conda creates the environment based on the Conda Lock. It is absolutely safe to do since all packages, including the dependencies, are in the Conda Lock anyway.
> > I have created an issue on pip so maybe it won't be needed in the future: [pypa/pip#10002](https://github.com/pypa/pip/issues/10002)
> 
> We should **not** have any packages which use the `@` symbol in their `install_requires` as we determined that functionality is fundamentally broken. Instead the `install_requires` should list the package name and then the `requirements.txt` should specify how/where to get the package from (IE a `git+http://` URL).

From what I saw working with `pip` on this action I can only agree but can you show me some example of this cooperation between `install_requires` and the `requirements.txt`?

ajelinski commented:
The Python scripts had many changes, e.g. docstrings were added, since I found many parts not easily understandable after these few months.

ajelinski commented:
@mithro Should I make any changes to have this PR merged?

mithro commented:
Check with @acomodi and @kgugala - If they are happy merge away!

acomodi commented:
@ajelinski LGTM

mithro commented:
@acomodi - Now we just need to deploy this everywhere?

acomodi commented:
@mithro Yes, we can test this right away with arch-defs for starters

mithro commented:
Would symbiflow-examples be an easier starting repo?
